File: frontend/src/types/broker.ts
// Broker bağlantı yönetimi tipleri

export type BrokerType = "paper" | "is_yatirim" | "gedik" | "deniz" | "garanti";
export type BrokerStatus = "connected" | "disconnected" | "error" | "pending";

export interface BrokerConnection {
  id: string;
  broker_name: string;
  is_active: boolean;
  is_paper_trading: boolean;
  label: string;
  status: BrokerStatus;
  last_connected_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface BrokerConnectionCreate {
  broker_name: BrokerType;
  credentials: Record<string, string>;
  is_paper_trading?: boolean;
  label?: string;
}

export interface BrokerConnectionUpdate {
  credentials?: Record<string, string>;
  is_paper_trading?: boolean;
  is_active?: boolean;
  label?: string;
}

export interface BrokerConnectionListResponse {
  items: BrokerConnection[];
  total: number;
}

export interface BrokerTestResult {
  success: boolean;
  broker_name: string;
  message: string;
  latency_ms: number | null;
}

export interface BrokerQuoteResponse {
  symbol: string;
  price: number;
  bid: number;
  ask: number;
  volume: number;
  source: string;
}

export interface BrokerInfo {
  name: BrokerType;
  display_name: string;
  description: string;
  requires_credentials: boolean;
  credential_fields: string[];
  is_available: boolean;
}

export interface BrokerListInfo {
  brokers: BrokerInfo[];
}

File: backend/app/schemas/broker.py
"""Broker bağlantı yönetimi şemaları."""

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field


class BrokerType(str, Enum):
    PAPER = "paper"
    IS_YATIRIM = "is_yatirim"
    GEDIK = "gedik"
    DENIZ = "deniz"
    GARANTI = "garanti"


class BrokerStatus(str, Enum):
    CONNECTED = "connected"
    DISCONNECTED = "disconnected"
    ERROR = "error"
    PENDING = "pending"


# ── Request Schemas ──


class BrokerConnectionCreate(BaseModel):
    broker_name: BrokerType = Field(..., description="Broker tipi")
    credentials: dict = Field(
        ...,
        description="Broker kimlik bilgileri (API key, username, password vb.)",
        examples=[{"api_key": "xxx", "api_secret": "yyy"}],
    )
    is_paper_trading: bool = Field(
        default=False,
        description="Paper trading modunda mı çalışacak",
    )
    label: str = Field(
        default="",
        max_length=100,
        description="Bağlantı için kullanıcı tanımlı etiket",
    )


class BrokerConnectionUpdate(BaseModel):
    credentials: dict | None = Field(None, description="Güncellenmiş kimlik bilgileri")
    is_paper_trading: bool | None = None
    is_active: bool | None = None
    label: str | None = Field(None, max_length=100)


# ── Response Schemas ──


class BrokerConnectionResponse(BaseModel):
    id: UUID
    broker_name: str
    is_active: bool
    is_paper_trading: bool
    label: str
    status: BrokerStatus
    last_connected_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BrokerConnectionListResponse(BaseModel):
    items: list[BrokerConnectionResponse]
    total: int


class BrokerTestResult(BaseModel):
    success: bool
    broker_name: str
    message: str
    latency_ms: float | None = None


class BrokerQuoteResponse(BaseModel):
    symbol: str
    price: float
    bid: float
    ask: float
    volume: int
    source: str = "paper"


# ── Broker Bilgileri ──


class BrokerInfo(BaseModel):
    name: BrokerType
    display_name: str
    description: str
    requires_credentials: bool
    credential_fields: list[str]
    is_available: bool


class BrokerListInfo(BaseModel):
    brokers: list[BrokerInfo]


# ── Yapılandırılmış broker bilgileri ──

BROKER_REGISTRY: dict[str, BrokerInfo] = {
    "paper": BrokerInfo(
        name=BrokerType.PAPER,
        display_name="Paper Trading",
        description="Sanal para ile simülasyon — gerçek emir gönderilmez",
        requires_credentials=False,
        credential_fields=[],
        is_available=True,
    ),
    "is_yatirim": BrokerInfo(
        name=BrokerType.IS_YATIRIM,
        display_name="İş Yatırım",
        description="İş Yatırım Menkul Değerler API entegrasyonu",
        requires_credentials=True,
        credential_fields=["api_key", "api_secret", "account_id"],
        is_available=False,
    ),
    "gedik": BrokerInfo(
        name=BrokerType.GEDIK,
        display_name="Gedik Yatırım",
        description="Gedik Yatırım Menkul Değerler API entegrasyonu",
        requires_credentials=True,
        credential_fields=["username", "password", "account_no"],
        is_available=False,
    ),
    "deniz": BrokerInfo(
        name=BrokerType.DENIZ,
        display_name="Deniz Yatırım",
        description="Deniz Yatırım Menkul Değerler API entegrasyonu",
        requires_credentials=True,
        credential_fields=["api_key", "secret_key"],
        is_available=False,
    ),
    "garanti": BrokerInfo(
        name=BrokerType.GARANTI,
        display_name="Garanti BBVA Yatırım",
        description="Garanti BBVA Yatırım API entegrasyonu",
        requires_credentials=True,
        credential_fields=["client_id", "client_secret", "account_id"],
        is_available=False,
    ),
}
